### PR TITLE
Assign shards and replicas to new classes based on the amount of free disk space

### PR DESCRIPTION
**Pull request by redouan-rhazouani**
### What's being changed:
This feature is related to new created classes
Nodes with the highest amount of free disk free spaces are selected as shards and replicas.
It uses the gossip protocol to spread the disk info into whole cluster
Two nodes are considered equivalent if space difference is less than 4KB

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: usecases/cluster/delegate.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+	d.set(d.Name, NodeInfo{space, time.Now().UnixMilli()}) // cache
+	return nil
+}
+
+// NodeMeta is used to retrieve meta-data about the current node
+// when broadcasting an alive message. It's length is limited to
+// the given byte size. This metadata is available in the Node structure.
+func (d *delegate) NodeMeta(limit int) (meta []byte) {
+	return nil
```
Comment (parkerduckworth) [inline usecases/cluster/delegate.go:117]: Is this intended to be a NOOP?

File: usecases/cluster/delegate.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	prv, _ := d.get(d.Name) // value from cache
+	var (
+		info NodeInfo = prv
+		err  error
+	)
+	// renew cached value if ttl expires
+	if prv.Available == 0 || time.Since(time.UnixMilli(prv.LastTimeMilli)) >= _ProtoTTL {
+		info.DiskUsage, err = d.diskUsage(d.dataPath)
+		if err != nil {
+			return nil
```
Comment (parkerduckworth) [inline usecases/cluster/delegate.go:134]: Would be a good idea to log this error

File: usecases/cluster/delegate.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	x := spaceMsg{
+		header{OpCode: _OpCodeDisk, ProtoVersion: _ProtoVersion},
+		info.DiskUsage,
+		d.Name,
+	}
+	bytes, err := x.marshal()
+	if err != nil {
+		return nil
```
Comment (parkerduckworth) [inline usecases/cluster/delegate.go:150]: Would be good to log this error

File: usecases/cluster/delegate.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+	}
+	info := NodeInfo{x.DiskUsage, time.Now().UnixMilli()}
+	d.set(x.Node, info)
+}
+
+func (d *delegate) NotifyMsg(data []byte) {}
+
+func (d *delegate) GetBroadcasts(overhead, limit int) [][]byte {
+	return nil
+}
```
Comment (parkerduckworth) [inline usecases/cluster/delegate.go:172]: Are `NotifyMsg`, `GetBroadcasts` intended to be NOOPs?

File: usecases/cluster/delegate.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+// get returns info about about a specific node in the cluster
+func (d *delegate) get(node string) (NodeInfo, bool) {
+	d.Lock()
+	defer d.Unlock()
```
Comment (parkerduckworth) [inline usecases/cluster/delegate.go:177]: Can we use `RWMutex` for `delegate` so we can `RLock` here instead?

File: usecases/cluster/delegate.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+	delete(d.Cache, node)
+}
+
+// sortCandidates by the amount of free space in descending order
+//
+// Two nodes are considered equivalent if the difference between their
+// free spaces is less than 4KB.
+// The free space is just an rough estimate of the actual amount.
+// The Lower bound 4KB helps to mitigate the risk of selecting same set of nodes
+// when selections happens concurrently on different initiator nodes.
```
Comment (parkerduckworth) [inline usecases/cluster/delegate.go:201]:
```
// The Lower bound 4KB helps to mitigate the risk of selecting same set of nodes
// when selections happens concurrently on different initiator nodes.
```
This is really nice 👍

File: usecases/cluster/delegate.go (hunk 4 above)
Comment (redouan-rhazouani) [reply usecases/cluster/delegate.go:172]: Yes they are (see explanation above regarding NodeMeta()

File: usecases/cluster/delegate.go (hunk 5 above)
Comment (redouan-rhazouani) [reply usecases/cluster/delegate.go:177]: In terms of performance Mutex is better. RWMutex makes more sense when you have large number of reader

File: usecases/cluster/delegate.go (hunk 2 above)
Comment (redouan-rhazouani) [reply usecases/cluster/delegate.go:134]: Yes it would 👍

File: usecases/cluster/delegate.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+// MergeRemoteState is invoked after a TCP Push/Pull. This is the
+// state received from the remote side and is the result of the
+// remote side's LocalState call. The 'join'
+// boolean indicates this is for a join instead of a push/pull.
+func (d *delegate) MergeRemoteState(data []byte, join bool) {
+	var x spaceMsg
+	if err := x.unmarshal(data); err != nil || x.Node == "" {
+		return
```
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply usecases/cluster/delegate.go:162]: Will do 👍

File: usecases/cluster/delegate.go (hunk 6 above)
Comment (redouan-rhazouani) [reply usecases/cluster/delegate.go:207]: Done

File: usecases/cluster/delegate.go (hunk 3 above)
Comment (redouan-rhazouani) [reply usecases/cluster/delegate.go:154]: Done

File: usecases/cluster/delegate.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply usecases/cluster/delegate.go:168]: Done

File: usecases/scaler/scaler.go [hunk 8]
```diff
@@ -71,7 +71,7 @@ type BackUpper interface {
 // cluster is used by the scaler to query cluster
 type cluster interface {
 	// AllNames returns list of existing node in the cluster
```
Comment (antas-marcin) [inline usecases/scaler/scaler.go:73]: Would be good also to change the comment above instead of `AllNames` should be `Candidates`

Comment (redouan-rhazouani) [reply usecases/scaler/scaler.go]: Done 👍

**Review by parkerduckworth — APPROVED** [review]

**Pull request merged**